cpp/src_prims/metrics/randIndex.h

```diff
@@ -0,0 +1,144 @@
+  uint64_t i = threadIdx.y + blockIdx.y*blockDim.y;
+
+  //thread-local variables to count a and b
+  uint64_t myA = 0, myB = 0;
+
+  if(i < size && j < size && j<i) {
+
+  //checking if the pair have been classified the same by both the clusters
+  if(firstClusterArray[i]==firstClusterArray[j]&&secondClusterArray[i]==secondClusterArray[j]){
+    ++myA;
+  }
+
+  //checking if the pair have been classified differently by both the clusters
+  else if(firstClusterArray[i]!=firstClusterArray[j]&&secondClusterArray[i]!=secondClusterArray[j]){
+    ++myB;
+    }
+  }
+
+  //specialize blockReduce for a 2D block of 1024 threads of type uint64_t
+  typedef cub::BlockReduce <uint64_t, BLOCK_DIM_X, cub::BLOCK_REDUCE_WARP_REDUCTIONS, BLOCK_DIM_Y> BlockReduce;
+
+  //Allocate shared memory for blockReduce
+  __shared__ typename BlockReduce::TempStorage temp_storage;
+  
+  //summing up thread-local counts specific to a block
+  myA = BlockReduce(temp_storage).Sum(myA);
+  __syncthreads();
+  myB = BlockReduce(temp_storage).Sum(myB);
+  __syncthreads();
+  
+  //executed once per block
+  if (threadIdx.x == 0 && threadIdx.y == 0) {
+    atomicAdd((unsigned long long int*)a, myA);
+    atomicAdd((unsigned long long int*)b, myB);
+  }
+}
+
+
+/**
+* @brief Function to calculate RandIndex
+* @param firstClusterArray: the array of classes of type T
+* @param secondClusterArray: the array of classes of type T
+* @param size: the size of the data points of type uint64_t
+* @param allocator: object that takes care of temporary device memory allocation of type std::shared_ptr<MLCommon::deviceAllocator>
+* @param stream: the cudaStream object
+*/
+template <typename T>
+float computeRandIndex (T* firstClusterArray, T* secondClusterArray, uint64_t size,
+                       std::shared_ptr<MLCommon::deviceAllocator> allocator, cudaStream_t stream) {
+
+  //allocating and initializing memory for a and b in the GPU
+  MLCommon::device_buffer<uint64_t> arr_buf (allocator, stream, 2);
+  CUDA_CHECK(cudaMemsetAsync(arr_buf.data(),0,2*sizeof(uint64_t),stream));
+
+  //kernel configuration
+  static const int BLOCK_DIM_Y = 16, BLOCK_DIM_X = 16;
+  dim3 numThreadsPerBlock(BLOCK_DIM_X, BLOCK_DIM_Y);
+  dim3 numBlocks(ceildiv<int>(size,numThreadsPerBlock.x),ceildiv<int>(size,numThreadsPerBlock.y));
+
+  //calling the kernel
+  computeTheNumerator<T, BLOCK_DIM_X, BLOCK_DIM_Y><<<numBlocks,numThreadsPerBlock,0,stream>>>(firstClusterArray,secondClusterArray,size,arr_buf.data(),arr_buf.data()+1);
+
+  //synchronizing and updating the calculated values of a and b from device to host
+  uint64_t ab_host[2] = {0};
+  MLCommon::updateHost(ab_host, arr_buf.data(), 2, stream);
+  CUDA_CHECK(cudaStreamSynchronize(stream));
+
+  //error handling
+  CUDA_CHECK(cudaGetLastError());
+
+  //denominator
+  uint64_t nChooseTwo = size*(size-1)/2;
+
+  //calculating the randIndex
+  return (float)(((float)(ab_host[0] + ab_host[1]))/(float)nChooseTwo);
```

Comment:
Since you are using `uint64_t` for computing `a` and `b`, better to return `double`, to minimize floating-point-induced inaccuracies.